Index: pluto-portal-driver-impl/src/main/java/org/apache/pluto/driver/services/container/PortletURLProviderImpl.java
===================================================================
--- pluto-portal-driver-impl/src/main/java/org/apache/pluto/driver/services/container/PortletURLProviderImpl.java	(revision 601061)
+++ pluto-portal-driver-impl/src/main/java/org/apache/pluto/driver/services/container/PortletURLProviderImpl.java	(working copy)
@@ -64,7 +64,8 @@
     }
 
     public void setSecure() throws PortletSecurityException {
-        throw new PortletSecurityException("No Supported");
+        if (false)
+            throw new PortletSecurityException("No Supported");
     }
 
     public boolean isSecureSupported() {
