Index: activemq-core/src/main/java/org/apache/activemq/transport/tcp/TcpTransportServer.java
===================================================================
--- activemq-core/src/main/java/org/apache/activemq/transport/tcp/TcpTransportServer.java	(revision 741739)
+++ activemq-core/src/main/java/org/apache/activemq/transport/tcp/TcpTransportServer.java	(working copy)
@@ -76,12 +76,7 @@
         InetAddress addr = InetAddress.getByName(host);
 
         try {
-            if (host.trim().equals("localhost") || addr.equals(InetAddress.getLocalHost())) {
-                this.serverSocket = serverSocketFactory.createServerSocket(bind.getPort(), backlog);
-            }
-            else {
-                this.serverSocket = serverSocketFactory.createServerSocket(bind.getPort(), backlog, addr);
-            }
+            this.serverSocket = serverSocketFactory.createServerSocket(bind.getPort(), backlog, addr);
             this.serverSocket.setSoTimeout(2000);
         }
         catch (IOException e) {
